### PR TITLE
Ignore future dated documents in hidden collection

### LINKED ISSUE #5953: Collection documents with future date and output set to false are not ignored
- [x] I believe this to be a bug, not a question about using Jekyll.
- [x] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [x] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [ ] This is a feature request.

---

- [ ] I am on (or have tested on) ***macOS*** 10+
- [x] I am on (or have tested on) ***Debian/Ubuntu*** GNU/Linux (Linux Mint)
- [ ] I am on (or have tested on) ***Fedora*** GNU/Linux
- [ ] I am on (or have tested on) ***Arch*** GNU/Linux
- [x] I am on (or have tested on) ***Other*** GNU/Linux
- [ ] I am on (or have tested on) ***Windows*** 10+

---

- [ ] I was trying to install.
- [ ] There is a broken Plugin API.
- [ ] I had an error on GitHub Pages, and I have reproduced it locally.
- [ ] I had an error on GitHub Pages, and GitHub Support said it was a Jekyll Bug.
- [ ] I had an error on GitHub Pages and I did not test it locally.
- [x] I was trying to build.
- [x] It was another bug.

## My Reproduction Steps

1. I created a collection, we'll call it "mycoll", with the following configuration in `_config.yml`:

```
collections:
  mycoll:
    output: false
```

2. I added a few files to the `_mycoll` directory; one in the past, one in the present, and one in the future. Each one contains front matter similar to the following:

```
date: 2017-03-16 00:00:00-04
link: http://example.com/place
preview: example.png
```

3. I used the following to fetch the latest item on my page:

```
{% assign item = site.mycoll.last %}
```

This is where the problem begins. For posts and collections with `output` set to `true`, posts and documents with the `date` field set to a time in the future are not published by default, unless built with `--future` or `future: true` in the config file; this is documented behavior. However, if `output` is set to `false`, objects with future dates are still displayed. This is not documented behavior.

This behavior is reproducible on both Jekyll 3.4.2 and 3.3.0.

## The Output I Wanted

What I would expect to happen is that ALL collection documents with future dates would be ignored, not just collections with `output: true`. Either that, or the documentation should be updated to reflect the current behavior.

Build output:

```
Configuration file: _config.yml
Configuration file: _config_beta.yml
            Source: _site
       Destination: public
 Incremental build: disabled. Enable with --incremental
      Generating...
Generating Compass: /home/site/beta/_site/_compass => /home/site/beta/public/css
```

/cc @jekyll/build @jekyll/documentation

### PR DESCRIPTION
**Pull request by ashmaroli**
Resolves #5953 

It may not be the most elegant approach, but it resolves the issue while preserving other existing functionality..

/cc @jekyll/build

### CODE REVIEW

**Comment by ashmaroli** [comment]
As of `v3.6.2`, documents in an unrendered collection (`output: false`) can be iterated through in Liquid (including future dated documents). Based on how you perceive that situation, this can be considered as either an `enhancement` or a `bug-fix`

**Comment by pathawks** [comment]
So, if I have a collection of upcoming events, and I am displaying them in a sidebar on my site, this enhancement will break that?

**Comment by ashmaroli** [comment]
> So, if I have a collection of upcoming events, and I am displaying them in a sidebar on my site, this enhancement will break that?

When you say `I am displaying` that means you probably built/served your site with `--future` which is then fine as demonstrated by the cucumber test..

**Comment by pathawks** [comment]
Iterating through them in a sidebar does not necessitate each item being output to its own page.

I agree that the expexted behavior is currently ambiguous and needs to be fixed, but we should probably call this a major change so users can be aware that the behavior has fundamentally changed.

**Comment by ashmaroli** [comment]
> So, if I have a collection of upcoming events, and I am displaying them in a sidebar on my site, this enhancement will break that?

I just tested.. this is indeed breaking that behavior.. :disappointed:

**Comment by pathawks** [comment]
I think we might consider making `future` available as a per-collection option.

Perhaps I don’t want my site built with future blog posts, but _do_ want future events.

I think the current `future` option was added when posts were the only dated content. That is no longer the case.

**Comment by ashmaroli** [comment]
> but we should probably call this a major change so users can be aware that the behavior has fundamentally changed.

But according to the scenario you described, people would probably **love** being able to do that.. I think the sane way out here is to apply a `wont-fix` label to the associated issue and document things to remove ambiguity..

**Comment by ashmaroli** [comment]
Though its frustrating that behaviors are different for future-dated docs in a Rendered collection (`output; true`) vs Unrendered collection (`output: false`)

**Comment by ashmaroli** [comment]
### Behavioral pattern for dated documents in a Rendered Collection vs Unrendered Collection

#### Default Jekyll build

<table>
<tr>
<th></th>
<th>Configuration</th>
<th colspan="2">Physical files at URL</th>
<th colspan="2">Access via liquid</th>
</tr>
<tr>
<td></td>
<td></td>
<td>Future dated</td>
<td>Non-future dated</td>
<td>Future dated</td>
<td>Non-future dated</td>
</tr>
<tr>
<td>Rendered Collection</td>
<td><code>output: true</code></td>
<td>:x:</td>
<td>:white_check_mark:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
</tr>
<tr>
<td>Unrendered Collection</td>
<td><code>output: false</code></td>
<td>:x:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
</tr>
</table>

---

#### Jekyll Build with `--future`

<table>
<tr>
<th></th>
<th>Configuration</th>
<th colspan="2">Physical files at URL</th>
<th colspan="2">Access via liquid</th>
</tr>
<tr>
<td></td>
<td></td>
<td>Future dated</td>
<td>Non-future dated</td>
<td>Future dated</td>
<td>Non-future dated</td>
</tr>
<tr>
<td>Rendered Collection</td>
<td><code>output: true</code></td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
</tr>
<tr>
<td>Unrendered Collection</td>
<td><code>output: false</code></td>
<td>:x:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
</tr>
</table>

**Comment by ashmaroli** [comment]
With this PR, the above table becomes..:


#### Default Jekyll build

<table>
<tr>
<th></th>
<th>Configuration</th>
<th colspan="2">Physical files at URL</th>
<th colspan="2">Access via liquid</th>
</tr>
<tr>
<td></td>
<td></td>
<td>Future dated</td>
<td>Non-future dated</td>
<td>Future dated</td>
<td>Non-future dated</td>
</tr>
<tr>
<td>Rendered Collection</td>
<td><code>output: true</code></td>
<td>:x:</td>
<td>:white_check_mark:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
</tr>
<tr>
<td>Unrendered Collection</td>
<td><code>output: false</code></td>
<td>:x:</td>
<td>:x:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
</tr>
</table>

---

#### Jekyll Build with `--future`

<table>
<tr>
<th></th>
<th>Configuration</th>
<th colspan="2">Physical files at URL</th>
<th colspan="2">Access via liquid</th>
</tr>
<tr>
<td></td>
<td></td>
<td>Future dated</td>
<td>Non-future dated</td>
<td>Future dated</td>
<td>Non-future dated</td>
</tr>
<tr>
<td>Rendered Collection</td>
<td><code>output: true</code></td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
</tr>
<tr>
<td>Unrendered Collection</td>
<td><code>output: false</code></td>
<td>:x:</td>
<td>:x:</td>
<td>:white_check_mark:</td>
<td>:white_check_mark:</td>
</tr>
</table>

File: lib/jekyll/collection.rb [hunk 1]
```diff
@@ -208,7 +208,7 @@ def read_document(full_path)
       doc = Jekyll::Document.new(full_path, :site => site, :collection => self)
       doc.read
       if site.publisher.publish?(doc) || !write?
```
Comment (parkr) [inline lib/jekyll/collection.rb]: Why doesn't `publish?` check `hidden_in_the_future?`?

**Comment by pathawks** [comment]
@ashmaroli What are your thoughts regarding making `future` a per-collection option?

File: lib/jekyll/collection.rb (hunk 1 above)
Comment (ashmaroli) [reply lib/jekyll/collection.rb]: > Why doesn't `publish?` check `hidden_in_the_future?` ?

@parkr, `:publish?` does `check hidden_in_the_future?` but it gets overridden by the `|| !write?`
`publisher == false` **but** `!write? == true` and therefore, the expression returns `true`

I initially thought about removing the `|| !write?` but changed it for some reason I do not remember now..

**Comment by ashmaroli** [comment]
> What are your thoughts regarding making `future` a per-collection option?

Isn't that too a `breaking-change`? Sites already built with `--future` or `future: true` suddenly won't be able to access future dated docs in their collection(s)..

**Comment by mattr-** [comment]
If it's a breaking change, it's a breaking change and we can do the Ruby things and this thing and release Jekyll 4.0. :smiley:

The topic of Jekyll 4.0 has come up a lot recently. My position is that if there's a breaking change that we want to make, we shouldn't be shy about making it and bumping the major version number. The `github-pages` gem, for example, is up to v167. The bumping of the version number communicates intent with the release, per SemVer, and IMHO, that's more valuable than dancing around releasing certain changes because they'd require a major version bump. The only caveat is that we need to be **very** clear about any potential breakages we'd cause as part of our release notes and our messaging around the major version release. I don't see that as being a large problem though.

**Comment by ashmaroli** [comment]
> we can do the Ruby things and this thing and release Jekyll 4.0.

Parker had mentioned long ago that he doesn't plan on shipping a 4.0 in the near future.. And if it changes now, we need a proper plan.. a roadmap.. to ensure things are well laid-out, tested and documented..
For that we have an open `4.0 milestone`

**Pull request closed without merging**